src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -377,19 +380,43 @@ RouterModifyTaskForShardInterval(Query *originalQuery, ShardInterval *shardInter
 	{
 		RelationRestriction *restriction = lfirst(restrictionCell);
 		List *originalBaserestrictInfo = restriction->relOptInfo->baserestrictinfo;
+		Var *relationPartitionKey = PartitionColumn(restriction->relationId,
+													restriction->index);
+		Param *uninstantiatedParameter = NULL;
 
 		/*
-		 * We haven't added the quals if all participating tables are reference
-		 * tables. Thus, now skip instantiating them.
+		 * We don't need to add restriction to reference tables given that they are
+		 * already single sharded and always prune to that single shard.
 		 */
-		if (allReferenceTables)
+		if (!relationPartitionKey)
 		{
-			break;
+			continue;
 		}
 
-		originalBaserestrictInfo =
-			(List *) InstantiatePartitionQual((Node *) originalBaserestrictInfo,
-											  shardInterval);
+		hashedOpExpressions = HashedShardIntervalOpExpressions(shardInterval);
+		Assert(list_length(hashedOpExpressions) == 2);
+
+		/*
+		 * Here we check whether the planner knows an equality between the partition column
+		 * and the uninstantiated parameter. If such an equality exists, we simply add the
+		 * shard restrictions.
+		 */
+		uninstantiatedParameter = UninstantiatedParameterForColumn(relationPartitionKey);
+		if (exprs_known_equal(restriction->plannerInfo, (Node *) relationPartitionKey,
```

Comment:
In his suggestion, @anarazel asked for the following:
> Instead of iterating baserestrictinfo like we currently do, we need a
routine very similar to equivclass.c's exprs_known_equal().  With a
version that additionally checks that EquivalenceClass->ec_relids
contains the current to-be-pruned relation

But, as far as I can see, we could directly use `exprs_known_equal()` given that Partition key already includes the correct varno and thus ensures that it belongs to the relation that is to-be-pruned (via the `equal` call in the `exprs_known_equal ()`).

Do you see any problems with using `exprs_known_equal ()` ?